test/cpp/api/modules.cpp

```diff
@@ -572,6 +572,39 @@ TEST_F(ModulesTest, MaxUnpool1d) {
   ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 1, 5}));
 }
 
+TEST_F(ModulesTest, MaxUnpool2d) {
+  auto indices = torch::tensor({
+  {{{ 6,  8,  9},
+    {16, 18, 19},
+    {21, 23, 24}}},
+  {{{ 6,  8,  9},
+    {16, 18, 19},
+    {21, 23, 24}}}}, torch::kLong);
+  auto x = torch::tensor({
+  {{{ 6,  8,  9},
+    {16, 18, 19},
+    {21, 23, 24}}},
+  {{{31, 33, 34},
+    {41, 43, 44},
+    {46, 48, 49}}}}, torch::requires_grad());
+  auto model = MaxUnpool2d{MaxUnpool2dOptions(3).stride(2).padding(1)};
+  auto y = model->forward(x, indices);
+
+  ASSERT_EQ(y.dim(), 4);
+  ASSERT_TRUE(torch::allclose(y, torch::tensor(
+   {{{{ 0,  0,  0,  0,  0},
+      { 0,  6,  0,  8,  9},
+      { 0,  0,  0,  0,  0},
+      { 0, 16,  0, 18, 19},
+      { 0, 21,  0, 23, 24}}},
+    {{{ 0,  0,  0,  0,  0},
+      { 0, 31,  0, 33, 34},
+      { 0,  0,  0,  0,  0},
+      { 0, 41,  0, 43, 44},
+      { 0, 46,  0, 48, 49}}}} , torch::kFloat)));
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({2, 1, 5, 5}));
+}
```

Comment:
It would be great to add a MaxPool2d -> MaxUnpool2d test, such as:
```cpp
  {
    MaxPool2d pool(MaxPool2dOptions(2).stride(2));
    MaxUnpool2d unpool(MaxUnpool2dOptions(2).stride(2));
    auto input = torch::tensor({{{{ 1.,  2.,  3.,  4.},
                                  { 5.,  6.,  7.,  8.},
                                  { 9., 10., 11., 12.},
                                  {13., 14., 15., 16.}}}});
    torch::Tensor output, indices;
    std::tie(output, indices) = pool->forward_with_indices(input);
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices),
      torch::tensor({{{{  0.,   0.,   0.,   0.},
                       {  0.,   6.,   0.,   8.},
                       {  0.,   0.,   0.,   0.},
                       {  0.,  14.,   0.,  16.}}}})));
    ASSERT_TRUE(torch::allclose(
      unpool(output, indices, /*output_size=*/{1, 1, 5, 5}),
      torch::tensor({{{{  0.,   0.,   0.,   0.,   0.},
                       {  6.,   0.,   8.,   0.,   0.},
                       {  0.,   0.,   0.,  14.,   0.},
                       { 16.,   0.,   0.,   0.,   0.},
                       {  0.,   0.,   0.,   0.,   0.}}}})));
  }
```